Sources/ConsoleKit/Activity/ActivityIndicator.swift

```diff
@@ -41,34 +44,58 @@ public final class ActivityIndicator<A> where A: ActivityIndicatorType {
     /// Current state.
     private var state: ActivityIndicatorState
     
-    private var task: DispatchWorkItem?
-
+    /// The queue on which to handle timer events
+    private let queue: DispatchQueue
+    
+    /// We use a DispatchGroup as a synchronization mechanism for when the
+    /// dispatch timer is cancelled.
+    private let stopGroup: DispatchGroup
+    
+    /// The timer that drives this activity indicator's updates.
+    private var timer: DispatchSourceTimer
+    
     /// Creates a new `ActivityIndicator`. Use `ActivityIndicatorType.newActivity(for:)`.
-    init(activity: A, console: Console) {
+    init(activity: A, console: Console, targetQueue: DispatchQueue? = nil) {
         self.console = console
         self.state = .ready
         self.activity = activity
+        self.queue = DispatchQueue(label: "codes.vapor.consolekit.activityindicator", target: targetQueue)
+        self.timer = DispatchSource.makeTimerSource(flags: [], queue: self.queue)
+        self.stopGroup = DispatchGroup()
     }
 
     /// Starts the `ActivityIndicator`. Usually this means beginning the associated "loading" animation.
     ///
     /// Once started, `ActivityIndicator` will continue to redraw the `ActivityIndicatorType` at a fixed
     /// refresh rate passing `ActivityIndicatorState.active`.
-    public func start() {
-        let item = DispatchWorkItem {
-            var tick: UInt = 0
-            while true {
-                usleep(40_000)
-                if tick > 0 {
-                    self.console.popEphemeral()
-                }
-                self.console.pushEphemeral()
-                self.activity.outputActivityIndicator(to: self.console, state: .active(tick: tick))
-                tick = tick &+ 1
+    ///
+    /// - Parameters:
+    ///     - refreshRate: The time interval (specified in milliseconds) to use
+    ///                    when updating the activity.
+    public func start(refreshRate: Int = 40) {
+        self.timer.schedule(deadline: DispatchTime.now(),
+                            repeating: .milliseconds(refreshRate),
```

Comment:
This type of formatting is hard to re-create on non-Apple platforms. I'd recommend using one indentation + new line:

```swift
self.timer.schedule(
    a: a,
    b: b
)
```

Reply:
@tanner0101 Fixed. Let me know if there is anything else I should clean up.